src/pkg/packager/publish.go

```diff
@@ -0,0 +1,369 @@
+		message.Debug("layer", desc.Digest.Hex()[:12], "exists")
+		return nil
+	}
+	copyOpts.PostCopy = func(ctx context.Context, desc ocispec.Descriptor) error {
+		message.SuccessF(desc.Digest.Hex()[:12])
+		return nil
+	}
+
+	push := func(root ocispec.Descriptor) error {
+		message.Debugf("root descriptor: %v\n", root)
+		if tag := dst.Reference.Reference; tag == "" {
+			err = oras.CopyGraph(ctx, store, dst, root, copyOpts.CopyGraphOptions)
+		} else {
+			_, err = oras.Copy(ctx, store, root.Digest.String(), dst, tag, copyOpts)
+		}
+		return err
+	}
+
+	// first attempt to do a ArtifactManifest push
+	root, err := pack()
+	if err != nil {
+		return err
+	}
+
+	copyRootAttempted := false
+	preCopy := copyOpts.PreCopy
+	copyOpts.PreCopy = func(ctx context.Context, desc ocispec.Descriptor) error {
+		message.Infof(desc.Digest.Hex()[:12])
+		if content.Equal(root, desc) {
+			// copyRootAttempted helps track whether the returned error is
+			// generated from copying root.
+			copyRootAttempted = true
+		}
+		if preCopy != nil {
+			return preCopy(ctx, desc)
+		}
+		return nil
+	}
+
+	// attempt to push the artifact manifest
+	if err = push(root); err == nil {
+		spinner.Updatef("Published: %s [%s]", ref, root.MediaType)
+		message.SuccessF("Published: %s [%s]", ref, root.MediaType)
+		message.SuccessF("Digest: %s", root.Digest)
+		return nil
+	}
+	// log the error, the expected error is a 400 manifest invalid
+	message.Debug(err)
+
+	if !copyRootAttempted || root.MediaType != ocispec.MediaTypeArtifactManifest ||
+		!isManifestUnsupported(err) {
+		return fmt.Errorf(`failed to push artifact manifest, 
+		was it during the copying of root? (%t)
+		was the root mediaType an artifact manifest? (%t)
+		was it because the registry does not support the artifact manifest mediaType? (%t)
+		
+		%w`, !copyRootAttempted, root.MediaType == ocispec.MediaTypeArtifactManifest, !isManifestUnsupported(err), err)
+	}
+
+	// assumes referrers API is not supported since OCI artifact
+	// media type is not supported
+	dst.SetReferrersCapability(false)
+
+	// fallback to an ImageManifest push
+	packOpts.PackImageManifest = true
+	root, err = pack()
+	if err != nil {
+		return err
+	}
+
+	copyOpts.FindSuccessors = func(ctx context.Context, fetcher content.Fetcher, node ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		if content.Equal(node, root) {
+			// skip non-config
+			content, err := content.FetchAll(ctx, fetcher, root)
+			if err != nil {
+				return nil, err
+			}
+			var manifest ocispec.Manifest
+			if err := json.Unmarshal(content, &manifest); err != nil {
+				return nil, err
+			}
+			return []ocispec.Descriptor{manifest.Config}, nil
+		}
+		// config has no successors
+		return nil, nil
+	}
+
+	if err = push(root); err != nil {
+		return err
+	}
+	spinner.Updatef("Published: %s [%s]", ref, root.MediaType)
+	message.SuccessF("Published: %s [%s]", ref, root.MediaType)
+	message.SuccessF("Digest: %s", root.Digest)
+	return nil
+}
```

Comment:
This function is several hundred lines long, I would recommend splitting it into at least private helper functions or if they are generic enough helper functions in the utils package.

Reply:
Did some breakup of the big `publish` fn. Will take another look tonight.